File: android/core/src/main/kotlin/com/metanav/core/ObstacleReasoner.kt
package com.metanav.core

import kotlin.math.max

/**
 * The whole pipeline after the neural nets: calibrate depth against the ground plane, extract
 * blobs, track them over time, label them with detections, and decide what (if anything) to say.
 * Pure Kotlin, no Android dependencies, deterministic given the inputs.
 */
class ObstacleReasoner(config: ReasonerConfig = ReasonerConfig()) {
    var config: ReasonerConfig = config
        set(value) {
            val geometryChanged = value.geometry != field.geometry
            field = value
            extractor = ObstacleExtractor(value)
            tracker = ObstacleTracker(value)
            policy = AdvisoryPolicy(value)
            if (geometryChanged) { groundCache = null; calibrator.reset() }
        }

    private val calibrator = DepthScaleCalibrator()
    private var extractor = ObstacleExtractor(config)
    private var tracker = ObstacleTracker(config)
    private var policy = AdvisoryPolicy(config)
    private var groundCache: GroundPlaneModel? = null
    private var lastScene = SceneState()

    val isCalibrated: Boolean get() = calibrator.isCalibrated

    fun reset() {
        calibrator.reset()
        tracker.reset()
        policy.reset()
        lastScene = SceneState()
    }

    fun process(observation: FrameObservation): ReasonerOutput {
        val depth = observation.depth ?: return ReasonerOutput(lastScene, null)
        val ground = groundFor(depth.height)
        calibrator.update(depth, ground)

        val extraction = extractor.extract(depth, calibrator, ground)
        val tracks = tracker.update(extraction.blobs)
        tracker.applyDetections(observation.detections, ground, depth.height)
        val frame = tracker.currentFrame()

        val confirmed = tracks.filter { it.isConfirmed(frame, config) && it.lastSeenFrame == frame }
        val inCorridor = confirmed.filter { it.overlapsCorridor(config) && it.distanceMeters < config.alertDistanceMeters }
        val candidate = inCorridor.minByOrNull { it.distanceMeters }

        val direction = if (candidate != null) chooseDirection(candidate, extraction) else Direction.NONE
        val advisory = policy.decide(observation.timestampMs, candidate, direction, corridorBlocked = candidate != null)

        val urgency = when {
            candidate == null -> Urgency.NONE
            candidate.distanceMeters < config.urgentDistanceMeters -> Urgency.STOP
            else -> Urgency.CAUTION
        }
        val nearest = confirmed.minOfOrNull { it.distanceMeters }
        val scene = SceneState(
            urgency = urgency,
            headline = when {
                !calibrator.isCalibrated -> "Calibrating"
                candidate == null -> "Clear"
                urgency == Urgency.STOP -> "Stop"
                else -> candidate.label
            },
            detail = when {
                !calibrator.isCalibrated -> "Point the camera down the path"
                candidate == null -> nearest?.let { "Nearest ${AdvisoryPolicy.shortDistance(it, config.units)}, off path" } ?: ""
                else -> "${AdvisoryPolicy.shortDistance(candidate.distanceMeters, config.units)} · ${directionText(direction)}"
            },
            direction = direction,
            nearestDistanceMeters = candidate?.distanceMeters ?: nearest,
            obstacles = confirmed.map {
                ObstacleInfo(it.id, it.label, it.distanceMeters, it.left, it.right, it.overlapsCorridor(config))
            },
            calibrated = calibrator.isCalibrated,
            columnDistances = extraction.columnDistances,
        )
        lastScene = scene
        return ReasonerOutput(scene, advisory)
    }

    private fun chooseDirection(candidate: Track, extraction: Extraction): Direction {
        val z = candidate.distanceMeters
        val leftFree = extraction.leftFreeMeters
        val rightFree = extraction.rightFreeMeters
        val minUseful = max(1.5f, z + 0.5f)
        val leftOk = leftFree >= minUseful
        val rightOk = rightFree >= minUseful
        // Prefer the side away from the obstacle's bulk; fall back to whichever side is freer.
        val awayFromObstacle = if (candidate.centerX > 0.5f) Direction.LEFT else Direction.RIGHT
        return when {
            leftOk && rightOk -> awayFromObstacle
            leftOk -> Direction.LEFT
            rightOk -> Direction.RIGHT
            else -> Direction.STOP
        }
    }

    private fun directionText(direction: Direction) = when (direction) {
        Direction.LEFT -> "move left"
        Direction.RIGHT -> "move right"
        Direction.STOP -> "stop"
        Direction.NONE -> ""
    }

    private fun groundFor(rows: Int): GroundPlaneModel {
        val cached = groundCache
        if (cached != null && cached.rows == rows) return cached
        return GroundPlaneModel(config.geometry, rows).also { groundCache = it }
    }
}

File: ios/MetanavCore/Sources/MetanavCore/ObstacleReasoner.swift
import Foundation

/// The whole pipeline after the neural nets: calibrate depth against the ground plane, extract
/// blobs, track them over time, label them with detections, and decide what (if anything) to say.
/// Pure Swift, no UIKit, deterministic given the inputs. Not thread-safe: call from one queue.
public final class ObstacleReasoner {
    public var config: ReasonerConfig {
        didSet {
            extractor = ObstacleExtractor(config: config)
            tracker = ObstacleTracker(config: config)
            policy = AdvisoryPolicy(config: config)
            if config.geometry != oldValue.geometry { groundCache = nil; calibrator.reset() }
        }
    }

    private let calibrator = DepthScaleCalibrator()
    private var extractor: ObstacleExtractor
    private var tracker: ObstacleTracker
    private var policy: AdvisoryPolicy
    private var groundCache: GroundPlaneModel?
    private var lastScene = SceneState()

    public init(config: ReasonerConfig = ReasonerConfig()) {
        self.config = config
        extractor = ObstacleExtractor(config: config)
        tracker = ObstacleTracker(config: config)
        policy = AdvisoryPolicy(config: config)
    }

    public var isCalibrated: Bool { calibrator.isCalibrated }

    public func reset() {
        calibrator.reset()
        tracker.reset()
        policy.reset()
        lastScene = SceneState()
    }

    public func process(_ observation: FrameObservation) -> ReasonerOutput {
        guard let depth = observation.depth else { return ReasonerOutput(scene: lastScene, advisory: nil) }
        let ground = groundFor(rows: depth.height)
        calibrator.update(depth: depth, ground: ground)

        let extraction = extractor.extract(depth: depth, calibrator: calibrator, ground: ground)
        let tracks = tracker.update(blobs: extraction.blobs)
        tracker.applyDetections(observation.detections, ground: ground, frameHeightPx: depth.height)
        let frame = tracker.currentFrame()

        let confirmed = tracks.filter { $0.isConfirmed(frameIndex: frame, config: config) && $0.lastSeenFrame == frame }
        let inCorridor = confirmed.filter { $0.overlapsCorridor(config) && $0.distanceMeters < config.alertDistanceMeters }
        let candidate = inCorridor.min { $0.distanceMeters < $1.distanceMeters }

        let direction: Direction = candidate.map { chooseDirection($0, extraction) } ?? .none
        let advisory = policy.decide(nowMs: observation.timestampMs, candidate: candidate, direction: direction, corridorBlocked: candidate != nil)

        let urgency: Urgency
        if let candidate {
            urgency = candidate.distanceMeters < config.urgentDistanceMeters ? .stop : .caution
        } else {
            urgency = .none
        }
        let nearest = confirmed.map(\.distanceMeters).min()

        var scene = SceneState()
        scene.urgency = urgency
        if !calibrator.isCalibrated {
            scene.headline = "Calibrating"
            scene.detail = "Point the camera down the path"
        } else if let candidate {
            scene.headline = urgency == .stop ? "Stop" : candidate.label
            scene.detail = "\(AdvisoryPolicy.shortDistance(candidate.distanceMeters, units: config.units)) · \(directionText(direction))"
        } else {
            scene.headline = "Clear"
            scene.detail = nearest.map { "Nearest \(AdvisoryPolicy.shortDistance($0, units: config.units)), off path" } ?? ""
        }
        scene.direction = direction
        scene.nearestDistanceMeters = candidate?.distanceMeters ?? nearest
        scene.obstacles = confirmed.map {
            ObstacleInfo(id: $0.id, label: $0.label, distanceMeters: $0.distanceMeters, left: $0.left, right: $0.right, inCorridor: $0.overlapsCorridor(config))
        }
        scene.calibrated = calibrator.isCalibrated
        scene.columnDistances = extraction.columnDistances
        lastScene = scene
        return ReasonerOutput(scene: scene, advisory: advisory)
    }

    private func chooseDirection(_ candidate: Track, _ extraction: Extraction) -> Direction {
        let z = candidate.distanceMeters
        let minUseful = max(1.5, z + 0.5)
        let leftOk = extraction.leftFreeMeters >= minUseful
        let rightOk = extraction.rightFreeMeters >= minUseful
        // Prefer the side away from the obstacle's bulk; fall back to whichever side is freer.
        let awayFromObstacle: Direction = candidate.centerX > 0.5 ? .left : .right
        switch (leftOk, rightOk) {
        case (true, true): return awayFromObstacle
        case (true, false): return .left
        case (false, true): return .right
        default: return .stop
        }
    }

    private func directionText(_ direction: Direction) -> String {
        switch direction {
        case .left: return "move left"
        case .right: return "move right"
        case .stop: return "stop"
        case .none: return ""
        }
    }

    private func groundFor(rows: Int) -> GroundPlaneModel {
        if let cached = groundCache, cached.rows == rows { return cached }
        let model = GroundPlaneModel(geometry: config.geometry, rows: rows)
        groundCache = model
        return model
    }
}
